Lib/profiling/sampling/_heatmap_assets/heatmap.js

```diff
@@ -331,6 +349,403 @@ document.addEventListener('click', e => {
+        if (opcodeList.length > 0) {
+            opcodesHtml = `
+                <div class="span-tooltip-section">Opcodes:</div>
+                ${opcodeList.map(op => `<div class="span-tooltip-opcode">${op}</div>`).join('')}
+            `;
+        }
+    }
+
+    const tooltip = document.createElement('div');
+    tooltip.className = 'span-tooltip';
+    tooltip.innerHTML = `
+        <div class="span-tooltip-header ${hotnessClass}">${hotnessText}</div>
+        <div class="span-tooltip-row">
+            <span class="span-tooltip-label">Samples:</span>
+            <span class="span-tooltip-value${isHot ? ' highlight' : ''}">${samples.toLocaleString()}</span>
+        </div>
+        <div class="span-tooltip-row">
+            <span class="span-tooltip-label">% of line:</span>
+            <span class="span-tooltip-value">${pct}%</span>
+        </div>
+        ${opcodesHtml}
+    `;
+
+    document.body.appendChild(tooltip);
+    activeTooltip = tooltip;
+
+    // Position tooltip above the span
+    const rect = span.getBoundingClientRect();
+    const tooltipRect = tooltip.getBoundingClientRect();
+
+    let left = rect.left + (rect.width / 2) - (tooltipRect.width / 2);
+    let top = rect.top - tooltipRect.height - 8;
+
+    // Keep tooltip in viewport
+    if (left < 5) left = 5;
+    if (left + tooltipRect.width > window.innerWidth - 5) {
+        left = window.innerWidth - tooltipRect.width - 5;
+    }
+    if (top < 5) {
+        top = rect.bottom + 8; // Show below if no room above
+    }
+
+    tooltip.style.left = `${left + window.scrollX}px`;
+    tooltip.style.top = `${top + window.scrollY}px`;
+}
+
+/**
+ * Hide active tooltip
+ */
+function hideSpanTooltip() {
+    if (activeTooltip) {
+        activeTooltip.remove();
+        activeTooltip = null;
+    }
+}
+
+/**
+ * Initialize span tooltip handlers
+ */
+function initSpanTooltips() {
+    document.addEventListener('mouseover', (e) => {
+        const span = e.target.closest('.instr-span');
+        if (span && specViewEnabled) {
+            showSpanTooltip(span);
+        }
+    });
+
+    document.addEventListener('mouseout', (e) => {
+        const span = e.target.closest('.instr-span');
+        if (span) {
+            hideSpanTooltip();
+        }
+    });
+}
+
+function toggleSpecView() {
+    specViewEnabled = !specViewEnabled;
+    const lines = document.querySelectorAll('.code-line');
+
+    lines.forEach(line => {
+        if (specViewEnabled) {
+            const specColor = line.getAttribute('data-spec-color');
+            line.style.background = specColor || 'transparent';
+        } else {
+            const bgColor = colorMode === 'self'
+                ? line.getAttribute('data-self-color') || line.getAttribute('data-bg-color')
+                : line.getAttribute('data-cumulative-color') || line.getAttribute('data-bg-color');
+            line.style.background = bgColor || 'transparent';
+        }
+    });
```

Comment:
This doesn't restore line colours:

https://github.com/user-attachments/assets/54a7eed3-ff5a-4d9a-91f8-09a88faf7082

```suggestion
    if (specViewEnabled) {
        lines.forEach(line => {
            const specColor = line.getAttribute('data-spec-color');
            line.style.background = specColor || 'transparent';
        });
    } else {
        applyLineColors();
    }
```